### PR TITLE
nixos/mosquitto: fix escaping, types, allow multiple listeners

### PR DESCRIPTION
**Pull request by pennae**
###### Motivation for this change

the mosquitto module interpolates stuff into shell scripts without escaping, which is a bit sad when you try to configure a literal password that contains a `'`. also some types are weird, and there's no overwhelming need to restrict the module to only one listener.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/modules/services/networking/mosquitto.nix [hunk 1]
```diff
@@ -35,6 +35,29 @@ let
...
+        type = types.str;
+        description = ''
+          Address to listen on.
+        '';
+      };
+
+      port = mkOption {
+        default = port;
+        example = port;
+        type = types.int;
```
Comment (nagy) [inline nixos/modules/services/networking/mosquitto.nix:51]:
```suggestion
        type = types.port;
```
?

**Comment by pennae** [comment]
maybe the module would benefit from being rewritten completely? there's a lot it currently cannot do (like enabling client cert authentication on only one listener), and the password file generation is pretty wonky one way or other. we don't really validate the config either, even though we could.

**Comment by pennae** [comment]
took a good long look at rewriting the module. due to deficiencies in the mosquitto config parser we'd have to do *a lot* of validation in the module, requiring one an option definition for almost every option mosquitto has. if that sounds like a good idea we can do that, but honestly we'd rather not.

**Comment by peterhoeg** [comment]
> took a good long look at rewriting the module. due to deficiencies in the mosquitto config parser we'd have to do a lot of validation in the module

I'm in favour on the minimum amount of work required to get it going.

**Comment by pennae** [comment]
rewrote the module a little. only very lightly tested and just the first thing that actually ran the test, will probably change quite a few things still.

**Comment by pennae** [comment]
okay, this is the general idea now. is this acceptable?

not quite happy with having freeform settings as extra attributes per level instead of having them on the same level as the other options, but putting them on the same level makes validation hard. if there's a way to query all freeform options of a submodule validation would be easy, but haven't found anything of the sort :(

**Comment by peterhoeg** [comment]
You've put in a ton of effort @pennae - I need to take some time out to go through this properly and in detail to give it the attention it deserves. This will realistically only happen this weekend.

**Comment by pennae** [comment]
no rush, and no worries. if it's bad it's bad, effort or no :D

**Comment by pennae** [comment]
don't want to be pushy, but is there a chance we'll get a review this weekend?

**Comment by peterhoeg** [comment]
I plan to take a look this weekend.

**Comment by pennae** [comment]
thank you! :+1:

**Comment by pennae** [comment]
@peterhoeg heya, it's totally fine if you don't have the time to review! if that is so, is there someone else who could take that over?

**Comment by peterhoeg** [comment]
Hey, no, sorry, I haven't. And realistically it isn't going to happen any week night (during the week). It's not cool to leave you hanging when you've put in all this effort. I *will* get to it but yes if anybody else wants to take a stab, please do.

**Comment by pennae** [comment]
friendly ping

**Comment by pennae** [comment]
i guess the total lack of anything happening is a good sign that we should just abandon this.

**Review by tomberek — CHANGES_REQUESTED** [review]
Breaking changes would make release notes helpful. Goal would be to merge prior to the next release cycle.

**Comment by pennae** [comment]
added release notes, reran tests (all pass).

**Review by tomberek — CHANGES_REQUESTED** [review]
The home-assistant NixOS tests now do not evaluate.

**Comment by pennae** [comment]
rebased, fixed home-assistant test to use the new config format. succeeds here :)

**Review by tomberek — APPROVED** [review]

**Pull request merged**